// fountain-js/dist.esm/token.js
import { rules } from './rules';
export class TitlePageBlock {
    constructor(source) {
        this.tokens = [];
        const titlePageBlock = /^\s*(?:[\w ]+(?<!\\)\:[^\S\n]*(?:(?:\n(?: {3}|\t))?[^\S\n]*\S.*)+(?:\n|$))+/;
        const keyValuePair = /^\s*[\w ]+(?<!\\)\:[^\S\n]*(?:(?:\n(?![\w ]+\:)(?: {3}|\t))?[^\S\n]*\S.*)+(?:\n|$)/;
        let scanPosition = 0;
        const match = source.match(titlePageBlock);
        if (match) {
            let titlePageData = match[0];
            this.scanIndex = titlePageData.length;
            while (scanPosition < this.scanIndex) {
                const pair = titlePageData.match(keyValuePair);
                if (pair) {
                    this.tokens = new TitlePageToken(pair[0]).addTo(this.tokens);
                    titlePageData = titlePageData.substring(pair[0].length);
                    scanPosition += pair[0].length;
                }
                else {
                    this.scanIndex = scanPosition;
                    break;
                }
            }
        }
    }
    addTo(tokens) {
        return [...tokens, ...this.tokens];
    }
    static matchedBy(source) {
        return rules.title_page.test(source);
    }
}
export class TitlePageToken {
    constructor(pair) {
        this.is_title = true;
        const [key, delimeter, value] = pair.split(/(\:[^\S\n]*\n?)([\S\s]*)/, 3);
        this.type = key.trim().toLowerCase().replace(/ /g, '_');
        this.text = value.replace(/^\s*|\s*$/gm, '');
    }
    addTo(tokens) {
        return [...tokens, this];
    }
}
export class SceneHeadingToken {
    constructor(line) {
        this.type = 'scene_heading';
        const match = line.match(rules.scene_heading);
        if (match) {
            this.text = (match[1] || match[2]).trim();
        }
        const meta = this.text.match(rules.scene_number);
        if (meta) {
            this.scene_number = meta[1];
            this.text = this.text.replace(rules.scene_number, '');
        }
    }
    addTo(tokens) {
        return [...tokens, this];
    }
    static matchedBy(line) {
        return rules.scene_heading.test(line);
    }
}
export class CenteredToken {
    constructor(line) {
        this.type = 'centered';
        const match = line.match(rules.centered);
        if (match) {
            this.text = match[0].replace(/[^\S\n]*[><][^\S\n]*/g, '');
        }
    }
    addTo(tokens) {
        return [...tokens, this];
    }
    static matchedBy(line) {
        return rules.centered.test(line);
    }
}
export class TransitionToken {
    constructor(line) {
        this.type = 'transition';
        const match = line.match(rules.transition);
        if (match) {
            this.text = (match[1] || match[2]).trim();
        }
    }
    addTo(tokens) {
        return [...tokens, this];
    }
    static matchedBy(line) {
        return rules.transition.test(line);
    }
}
export class DialogueBlock {
    constructor(line, dual) {
        this.tokens = [];
        const match = line.match(rules.dialogue);
        if (match) {
            this.scanIndex = match.length;
            let name = match[1].trim();
            // iterating from the bottom up, so push dialogue blocks in reverse order
            const isDualDialogue = !!match[2];
            if (isDualDialogue) {
                this.tokens.push(new DualDialogueEndToken());
            }
            this.tokens.push(new DialogueEndToken());
            const parts = match[3].split(/\n/);
            let dialogue = parts.reduce((p, text = '') => {
                const lastIndex = p.length - 1;
                const previousToken = p[lastIndex];
                if (!text.length) {
                    return p;
                }
                if (rules.line_break.test(text)) {
                    text = '';
                }
                text = text.trim();
                if (rules.parenthetical.test(text)) {
                    return [...p, new ParentheticalToken(text)];
                }
                if (rules.lyrics.test(text)) {
                    if (previousToken.type === 'lyrics') {
                        p[lastIndex].text =
                            `${previousToken.text}\n${text.replace(/^~/, '')}`;
                        return p;
                    }
                    else {
                        return [...p, new LyricsToken(text)];
                    }
                }
                if ((previousToken === null || previousToken === void 0 ? void 0 : previousToken.type) === 'dialogue') {
                    p[lastIndex].text = `${previousToken.text}\n${text}`;
                    return p;
                }
                return [...p, new DialogueToken(text)];
            }, []).reverse();
            this.tokens.push(...dialogue);
            this.tokens.push(new CharacterToken(name.startsWith('@')
                ? name.replace(/^@/, '').trim()
                : name.trim()), new DialogueBeginToken(isDualDialogue ? 'right' : dual ? 'left' : undefined));
            if (dual) {
                this.tokens.push(new DualDialogueBeginToken());
            }
            this.dual = isDualDialogue;
        }
    }
    addTo(tokens) {
        return [...tokens, ...this.tokens];
    }
    static matchedBy(line) {
        return rules.dialogue.test(line);
    }
}
export class DialogueBeginToken {
    constructor(dual) {
        this.type = 'dialogue_begin';
        this.dual = dual;
    }
    addTo(tokens) {
        return [...tokens, this];
    }
}
export class CharacterToken {
    constructor(text) {
        this.type = 'character';
        this.text = text;
    }
    addTo(tokens) {
        return [...tokens, this];
    }
}
export class DialogueToken {
    constructor(text) {
        this.type = 'dialogue';
        this.text = text;
    }
    addTo(tokens) {
        return [...tokens, this];
    }
}
export class DialogueEndToken {
    constructor() {
        this.type = 'dialogue_end';
    }
    addTo(tokens) {
        return [...tokens, this];
    }
}
export class ParentheticalToken {
    constructor(text) {
        this.type = 'parenthetical';
        this.text = text;
    }
    addTo(tokens) {
        return [...tokens, this];
    }
}
export class DualDialogueBeginToken {
    constructor() {
        this.type = 'dual_dialogue_begin';
    }
    addTo(tokens) {
        return [...tokens, this];
    }
}
export class DualDialogueEndToken {
    constructor() {
        this.type = 'dual_dialogue_end';
    }
    addTo(tokens) {
        return [...tokens, this];
    }
}
export class LyricsToken {
    constructor(line) {
        this.type = 'lyrics';
        this.text = line.replace(/^\s*~(?! )/gm, '');
    }
    addTo(tokens) {
        return [...tokens, this];
    }
    static matchedBy(line) {
        return rules.lyrics.test(line);
    }
}
export class SectionToken {
    constructor(line) {
        this.type = 'section';
        const match = line.match(rules.section);
        if (match) {
            this.text = match[2];
            this.depth = match[1].length;
        }
    }
    addTo(tokens) {
        return [...tokens, this];
    }
    static matchedBy(line) {
        return rules.section.test(line);
    }
}
export class SynopsisToken {
    constructor(line) {
        this.type = 'synopsis';
        const match = line.match(rules.synopsis);
        if (match) {
            this.text = match[1];
        }
    }
    addTo(tokens) {
        return [...tokens, this];
    }
    static matchedBy(line) {
        return rules.synopsis.test(line);
    }
}
export class NoteToken {
    constructor(line) {
        this.type = 'note';
        const match = line.match(rules.note);
        if (match) {
            this.text = match[1];
        }
    }
    addTo(tokens) {
        return [...tokens, this];
    }
    static matchedBy(line) {
        return rules.note.test(line);
    }
}
export class PageBreakToken {
    constructor() {
        this.type = 'page_break';
    }
    addTo(tokens) {
        return [...tokens, this];
    }
    static matchedBy(line) {
        return rules.page_break.test(line);
    }
}
export class SpacesToken {
    constructor() {
        this.type = 'spaces';
    }
    addTo(tokens) {
        return [...tokens, this];
    }
    static matchedBy(line) {
        return rules.blank_lines.test(line);
    }
}
export class ActionToken {
    constructor(line) {
        this.type = 'action';
        this.text = line
            .replace(/^(\s*)!(?! )/gm, '$1')
            .replace(/^( *)(\t+)/gm, (_, leading, tabs) => {
            return leading + '    '.repeat(tabs.length);
        });
    }
    addTo(tokens) {
        return [...tokens, this];
    }
}
